src/server/src/main.rs

```diff
@@ -1,23 +1,215 @@
+    }
+}
+
+async fn update_user_handler(
+    req: HttpRequest,
+    path: web::types::Path<(String,)>,
+    update_value: web::types::Json<String>,
+    user_map: web::types::Data<UserMap>,
+    session_map: web::types::Data<SessionMap>,) -> impl Responder 
+{
+    let (username,) = path.into_inner();
+    let session_id = req.headers().get("SessionID").and_then(|val| val.to_str().ok());
+    let session_map = session_map.lock().unwrap();
+
+    if let Some(sess_user) = session_id.and_then(|sid| session_map.get(sid)) 
+    {
+        let mut user_map = user_map.lock().unwrap();
+        if let Some(user) = user_map.get_mut(&username) 
+        {
+            user.password = update_value.into_inner(); //whatever we want update to do, just pw rn
+            return HttpResponse::Ok().json(user); 
+        }
+    }
+    HttpResponse::build(403).body("Incorrect sessionId for user.")
+}
+
+async fn login_user_handler(
+    body: web::types::Json<User>,
+    path: web::types::Path<(String,)>,
+    user_map: web::types::Data<UserMap>,
+    session_map: web::types::Data<SessionMap>,) -> impl Responder 
+{
+    let login_user = body.into_inner();
+    let (username,) = path.into_inner();
+    let map = user_map.lock().unwrap();
+
+    if let Some(user) = map.get(&username) 
+    {
+        if user.password == login_user.password 
+        {
+            // generate new session ID -> spawns a child process for the user session -> stores the session ID and child pid
+            let session_id = Uuid::new_v4().to_string();
+            let session_process = spawn_user_session(username.clone());
+            let mut session_map = session_map.lock().unwrap();
+            session_map.insert(session_id.clone(), session_process);
+            return HttpResponse::Ok().json(session_id); //returns id token
+        }
+    }
+    HttpResponse::BadRequest().body("Invalid username/password combination.")
+}
+
+async fn logout_user_handler(
+    body: web::types::Json<String>,
+    path: web::types::Path<(String,)>,
+    session_map: web::types::Data<SessionMap>,) -> impl Responder 
+{
+    let session_id = body.into_inner();
+    let (username,) = path.into_inner();
+    let mut session_map = session_map.lock().unwrap();
+
+    if session_map.remove(&session_id).is_some() 
+    {
+        HttpResponse::Ok().body("User successfully logged out.")
+    } else 
+    {
+        HttpResponse::BadRequest().body("Invalid sessionId.")
+    }
 }
 
 #[ntex::main]
-async fn main() -> std::io::Result<()> {
-    web::HttpServer::new(|| web::App::new().service(index).service(hello))
-        .bind(("127.0.0.1", 8080))?
-        .run()
-        .await
-}
\ No newline at end of file
+async fn main() -> std::io::Result<()> 
+{
+    Bastion::init();
+    Bastion::start();
+
+    // sharing state
+    let user_map = web::types::Data::new(Arc::new(Mutex::new(HashMap::new())));
+    let session_map = web::types::Data::new(Arc::new(Mutex::new(HashMap::new())));
```

Comment:
there should probably be some thin abstraction over these data types.

Reply:
(getting rid of `user_map ` is where i come in with the DB, but the `session_map` should remain a thing in volatile memory).